cmd/scollector/collectors/systemd_linux.go

```diff
@@ -0,0 +1,137 @@
+package collectors
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	"bosun.org/_third_party/github.com/coreos/go-systemd/dbus"
+	"bosun.org/cmd/scollector/conf"
+	"bosun.org/metadata"
+	"bosun.org/opentsdb"
+	"bosun.org/util"
+)
+
+type systemdServiceConfig struct {
+	regex *regexp.Regexp
+	watch bool
+}
+
+var systemdServices []*systemdServiceConfig
+
+func init() {
+	// The following two func defs have no-op stubs in collectors.go, as they are
+	// called by main.go, but only functional on Linux environments.
+	WatchSystemd = func() {
+		collectors = append(collectors, &IntervalCollector{
+			F: func() (opentsdb.MultiDataPoint, error) {
+				return c_systemd()
+			},
+			name: "c_systemd",
+		})
+	}
+
+	AddSystemdServiceConfig = func(params conf.ServiceParams) error {
+		if params.Name == "" {
+			return fmt.Errorf("empty service Name")
+		}
+		reg, err := regexp.Compile(params.Name)
+		if err != nil {
+			return err
+		}
+		serviceConfig := systemdServiceConfig{regex: reg, watch: params.WatchProc}
+		systemdServices = append(systemdServices, &serviceConfig)
+		return nil
+	}
+}
+
+// c_systemd() iterates through all units provided by systemd's dbus info.
+// If the unit is a service we care about, it sends service metrics.
+// Also sends process data if WatchProc was set to true in the [[SystemdService]] config.
+func c_systemd() (opentsdb.MultiDataPoint, error) {
+	conn, err := dbus.New()
+	if err != nil {
+		return nil, err
+	}
+
+	units, err := conn.ListUnits()
+	if err != nil {
+		return nil, err
+	}
+
+	var md opentsdb.MultiDataPoint
+	for _, unit := range units {
+		if strings.HasSuffix(unit.Name, ".service") {
+			shortName := strings.TrimSuffix(unit.Name, ".service")
+			for _, config := range systemdServices {
+				if config.regex.MatchString(shortName) {
+					if config.watch {
+						err = watchSystemdServiceProc(&md, conn, unit)
+						if err != nil {
+							return nil, err
+						}
+					}
+					systemdTags := opentsdb.TagSet{"name": unit.Name}
+					osTags := opentsdb.TagSet{"name": shortName}
+					Add(&md, "linux.systemd.unit.activestate", activeState[unit.ActiveState], systemdTags, metadata.Gauge, metadata.StatusCode, descActiveState)
```

Comment:
If we get an unexpected state, it will register as active (0). Perhaps we should check before this and replace it with a higher value just to be safe?